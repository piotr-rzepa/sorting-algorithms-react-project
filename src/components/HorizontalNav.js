/*
 * Komponent bezklasowy będacy menu poziomym, zawierającym pole wyszukiwania
 */

import React from 'react';
import { ReactComponent as SearchSVG } from '../images/search.svg';
import { ReactComponent as GitHubSVG } from '../images/github.svg';
import { ReactComponent as InstagramSVG } from '../images/instagram.svg';
import { ReactComponent as FacebookSVG } from '../images/facebook.svg';
import { ReactComponent as LogoSVG } from '../images/logo-p.svg';
import { Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import selectSearchText from '../selectors/searchTextSelector';
import { setTextFilter } from '../actions/filters';

export default function HorizontalNav() {
	const dispatch = useDispatch();
	const filteredAlgorithms = useSelector((state) => selectSearchText(state));
	const handleTextChange = (e) => {
		dispatch(setTextFilter(e.target.value));
	};
	return (
		<div className="HorizontalNav">
			<div className="logoContainer">
				<LogoSVG width="24px" height="24px" className="svg-image"></LogoSVG>
			</div>
			<SearchSVG width="24px" height="24px" />
			<input
				type="search"
				placeholder="Search..."
				onChange={handleTextChange}
			></input>
			<ul>
				{filteredAlgorithms.map((algorithm) => (
					<React.Fragment key={algorithm.id}>
						<Link to={`/algorithm/${algorithm.name}`}>{algorithm.name}</Link>
					</React.Fragment>
				))}
			</ul>
			<p>Piotr Rzepkowski</p>
			<h5>Creator</h5>
			<Link to={{ pathname: 'https://github.com/piotr-rzepa' }} target="_blank">
				<GitHubSVG width="24px" height="24px" />
			</Link>
			<Link
				to={{ pathname: 'https://www.instagram.com/piotrullo_/' }}
				target="_blank"
			>
				<InstagramSVG width="24px" height="24px" />
			</Link>
			<Link
				to={{ pathname: 'https://www.facebook.com/piotr.rzepkowski.3' }}
				target="_blank"
			>
				<FacebookSVG width="24px" height="24px" />
			</Link>
		</div>
	);
}
