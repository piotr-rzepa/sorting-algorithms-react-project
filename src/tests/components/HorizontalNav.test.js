import * as React from 'react';
import { shallow } from 'enzyme';
import { Link } from 'react-router-dom';
import { ReactComponent as GitHubSVG } from '../../images/github.svg';
import { ReactComponent as InstagramSVG } from '../../images/instagram.svg';
import { ReactComponent as FacebookSVG } from '../../images/facebook.svg';
import { ReactComponent as LogoSVG } from '../../images/logo-p.svg';
import HorizontalNav from '../../components/HorizontalNav';
import * as reactRedux from 'react-redux';

const data = [
	{
		id: 0,
		name: 'Bubble Sort',
		description:
			'Bubble sort, sometimes referred to as sinking sort, is a simple sorting algorithm',
		method: 'Exchanging',
		sortingTime: { best: 0, worst: 0, average: 0 }
	},
	{
		id: 1,
		name: 'Quick Sort',
		description: 'Quicksort is a divide-and-conquer algorithm.',
		method: 'Partitioning',
		sortingTime: { best: 0, worst: 0, average: 0 }
	}
];

describe('<HorizontalNav />', () => {
	const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
	const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
	const useStateMock = jest.spyOn(React, 'useState');
	const useContextMock = jest.spyOn(React, 'useContext');
	beforeEach(() => {
		useDispatchMock.mockClear();
		useSelectorMock.mockClear();
		useStateMock.mockClear();
		useContextMock.mockClear();
	});

	it('Should render Horizontal navbar correctly', () => {
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useStateMock.mockImplementation((state) => [state, jest.fn()]);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<HorizontalNav />);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should render expected amount of children components', () => {
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		useStateMock.mockImplementation((state) => [state, jest.fn()]);
		const wrapper = shallow(<HorizontalNav />);
		expect(wrapper.find(GitHubSVG)).toHaveLength(1);
		expect(wrapper.find(InstagramSVG)).toHaveLength(1);
		expect(wrapper.find(FacebookSVG)).toHaveLength(1);
		expect(wrapper.find(LogoSVG)).toHaveLength(1);
		expect(wrapper.find(Link)).toHaveLength(6);
	});

	it('Should set filter correctly based on user input', () => {
		const setState = jest.fn((fn) => fn());
		useDispatchMock.mockReturnValue(() => 'sort');
		useSelectorMock.mockReturnValue(data);
		useStateMock.mockImplementation((state) => [state, setState]);
		const wrapper = shallow(<HorizontalNav />);
		wrapper
			.find('input')
			.at(0)
			.simulate('change', { target: { value: 'sort' } });
		expect(useDispatchMock).toHaveBeenCalled();
	});

	it('Should show dropdown correctly based on user input', () => {
		const setState = jest.fn((fn) => fn());
		useDispatchMock.mockReturnValue(() => 'sort');
		useSelectorMock.mockReturnValue(data);
		useStateMock.mockImplementation((state) => [state, setState]);
		const wrapper = shallow(<HorizontalNav />);
		wrapper
			.find('input')
			.at(0)
			.simulate('change', { target: { value: 'sort' } });
		expect(useDispatchMock).toHaveBeenCalled();
		expect(setState).toBeCalled();
		expect(setState.mock.calls[0][0]).toEqual(expect.any(Function));
		expect(setState.mock.results[0].type).toEqual('return');
		expect(setState.mock.results[0].value).toBeTruthy();
	});

	it('Should hide dropdown if no input is provided', () => {
		const setState = jest.fn((fn) => fn());
		useDispatchMock.mockReturnValue(() => 'sort');
		useSelectorMock.mockReturnValue(data);
		useStateMock.mockImplementation((state) => [state, setState]);
		const wrapper = shallow(<HorizontalNav />);
		wrapper
			.find('input')
			.at(0)
			.simulate('change', { target: { value: '' } });
		expect(useDispatchMock).toHaveBeenCalled();
		expect(setState).toBeCalled();
		expect(setState.mock.calls[0][0]).toEqual(expect.any(Function));
		expect(setState.mock.results[0].type).toEqual('return');
		expect(setState.mock.results[0].value).toBeFalsy();
	});

	it('should choose correct light theme', () => {
		const setStateMock = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useStateMock.mockReturnValue([true, setStateMock]);
		useDispatchMock.mockReturnValue(jest.fn());
		useContextMock.mockReturnValue(false);
		const wrapper = shallow(<HorizontalNav />);
		expect(wrapper.find('div.HorizontalNav').get(0).props.style).toHaveProperty(
			'color',
			'black'
		);
	});
	it('should choose correct black theme', () => {
		const setStateMock = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useStateMock.mockReturnValue([true, setStateMock]);
		useDispatchMock.mockReturnValue(jest.fn());
		useContextMock.mockReturnValue(true);
		const wrapper = shallow(<HorizontalNav />);
		expect(wrapper.find('div.HorizontalNav').get(0).props.style).toHaveProperty(
			'color',
			'white'
		);
	});
});
