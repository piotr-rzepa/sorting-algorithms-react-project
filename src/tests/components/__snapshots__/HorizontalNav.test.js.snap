// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HorizontalNav /> Should render Horizontal navbar correctly 1`] = `
<div
  className="HorizontalNav"
  style={
    Object {
      "backgroundColor": "white",
      "color": "black",
    }
  }
>
  <Link
    className="logoContainer"
    to="/"
  >
    <ForwardRef(SvgLogoP)
      className="svg-image"
      height="24px"
      width="24px"
    />
  </Link>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="Search..."
      style={
        Object {
          "backgroundColor": "white",
          "color": "black",
        }
      }
      type="search"
    />
    <div
      className="dropdown-search "
    >
      <ul>
        <a>
          <Link
            to="/algorithm/Bubble Sort"
          >
            Bubble Sort
          </Link>
          <hr
            style={
              Object {
                "borderTop": "gray",
              }
            }
          />
        </a>
        <a>
          <Link
            to="/algorithm/Quick Sort"
          >
            Quick Sort
          </Link>
          <hr
            style={
              Object {
                "borderTop": "gray",
              }
            }
          />
        </a>
      </ul>
    </div>
  </div>
  <div
    className="creator-data"
  >
    <div
      className="creator-data-wrapper"
    >
      <img
        alt="Not found"
        className="rounded-img"
        src="ASC_2746_v1.jpg"
      />
      <div
        className="creator-data-name"
      >
        <p>
          Piotr Rzepkowski
        </p>
        <h5>
          Creator
        </h5>
      </div>
    </div>
    <div
      className="creator-data-socials"
    >
      <Link
        target="_blank"
        to={
          Object {
            "pathname": "https://github.com/piotr-rzepa",
          }
        }
      >
        <ForwardRef(SvgGithub)
          height="24px"
          width="24px"
        />
      </Link>
      <Link
        target="_blank"
        to={
          Object {
            "pathname": "https://www.instagram.com/piotrullo_/",
          }
        }
      >
        <ForwardRef(SvgInstagram)
          height="24px"
          width="24px"
        />
      </Link>
      <Link
        target="_blank"
        to={
          Object {
            "pathname": "https://www.facebook.com/piotr.rzepkowski.3",
          }
        }
      >
        <ForwardRef(SvgFacebook)
          height="24px"
          width="24px"
        />
      </Link>
    </div>
  </div>
</div>
`;
